### ISSUE TITLE
Question on permissions

### ISSUE THREAD

**Issue by IzzySoft** [sergenious/MediaBrowserAndroid#3]
For the last update, the scanners at IzzyOnDroid just sent me a warning:

```
! repo/com.sergenious.mediabrowser_100100.apk declares sensitive permission(s): android.permission.MANAGE_EXTERNAL_STORAGE
! repo/com.sergenious.mediabrowser_100100.apk contains signature block blobs: 0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)
```

Could you please clarify what `MANAGE_EXTERNAL_STORAGE` is needed for by a media browser?

As for `DEPENDENCY_INFO_BLOCK`, that can easily be avoided with a minor addition to your `build.gradle`:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains. More details can be found e.g. here: [Ramping up security: additional APK checks are in place with the IzzyOnDroid repo](https://android.izzysoft.de/articles/named/iod-scan-apkchecks#blobs).

Thanks in advance!

**Comment by sergenious** [comment]
Android by default doesn't allow listing physical files on internal or external storages. Until lately, permission READ_EXTERNAL_STORAGE was enough to allow that.

However, in the latest Android versions, due to stupid Google's paranoia, this is not enough anymore. READ_EXTERNAL_STORAGE now only allows listing the files in some specific directories, like "DCIM". In order to browse all the files anywhere (except the internal protected storage), users need to accept MANAGE_EXTERNAL_STORAGE permission as well.

Google's recommendation in order to avoid that, is using content resolvers. However, this is completely stupid, because content resolver lists the files flat, without any directory info. And this is completely the opposite of the Media Browser's purpose, which is listing through the directories, like on the computer. This is also the reason why most of the other photo gallery apps display the photos as a flat monotonous list.

So, unfortunately, because of Google's shortsightness, MANAGE_EXTERNAL_STORAGE is needed to be accepted.

I made this app primarily for myself, because I have all my photos (about 60000) structured in directories, by decade, year, and the event, where the directory name contains the event date, location and the purpose. So I need a complete freedom to browse through those directories.

**Comment by sergenious** [comment]
In don't know about the DEPENDENCY_INFO_BLOCK, it was probably automatically added by Android Studio. Will use your recommendation and try if it works properly without it.

**Comment by IzzySoft** [comment]
> DEPENDENCY_INFO_BLOCK, it was probably automatically added by Android Studio

It is automatically added when signing via Studio, IntelliJ and some other tools for a while now by default (which wasn't the case before), yes.

> Will use your recommendation and try if it works properly without it.

It will. The only possible side-effect known to me is not affecting you: that Google might have their PlayProtect yelling in case you also use AccessibilityServices :man_shrugging: You don't use that, so no risk.

> Google's recommendation in order to avoid that, is

to :roll_eyes: and stop thinking or asking yourself if the will *ever* get the mess with storage permissions sorted. And there I thought the mess only existed with Android 5-9. So with 13 or 14 they decided to introduce another mess, wonderful. "Manage" for "Reading" sounds sooo intuitive… `</s>` And then they block your app from being listed in their toy shop because MANAGE is reserved for file managers, right? Oh my…

TL;DR: afraid it's close to impossible to remember all the crazy exceptions and "special rules" for which storage permission is needed for what purpose on which Android version…

No, _you_ cannot be blamed for that. The crazy side effect of the mess is I'm close to believing anything in this context, as nothing seems too weird here anymore when "manage" permissions are needed for "read only" access…

Added to the green list of your app:

```yaml
    android.permission.READ_MEDIA_IMAGES: needed to access the media
    android.permission.READ_EXTERNAL_STORAGE: needed to access the media
    android.permission.READ_MEDIA_AUDIO: needed to access the media
    android.permission.READ_MEDIA_VIDEO: needed to access the media
    android.permission.MANAGE_EXTERNAL_STORAGE: needed to list all files in hierarchical structure
```

Thanks!
